File: backend/app/api/event.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import func
from sqlalchemy.orm import Session

from datetime import date
import re

from app.database import get_db
from app.api.deps import get_current_user, get_current_admin
from app.models.models import Event, Match, Team
from app.schemas.event import EventRequest, EventResponse, EventsListResponse

router = APIRouter()


@router.get("", response_model=EventsListResponse)
def list_events(db: Session = Depends(get_db), _: str = Depends(get_current_user)):
    """
    This function gets all the events.

    param : db - The database.
    param : _ - The client.
    return : Return all the events.
    """
    events = db.query(Event).all()
    return EventsListResponse(events=events, total=len(events))



@router.get("/{event_id}", response_model=EventResponse)
def get_event(event_id: int, db: Session = Depends(get_db), _: str = Depends(get_current_user)):
    """
    This function gets a specific event.

    param : event_id - The event's id.
    param : db - The session of database.
    param : _ - The client.
    return : Return the event.
    """
    event = db.query(Event).get(event_id)
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
    return EventResponse.model_validate(event)



@router.post("", response_model=EventResponse, status_code=status.HTTP_201_CREATED)
def create_event(data: EventRequest, db: Session = Depends(get_db), _: str = Depends(get_current_admin)):
    """
    This function creates a event.

    param : data - The event's informations.
    param : db - The session of database.
    param : _ - The client.
    return : Return the event created.
    """
    if data.event_date < date.today():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Back to the future")
    
    if not re.match(r"^(?:[01]\d|2[0-3]):[0-5]\d$", data.event_time):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="event_time must be in HH:MM format (00:00–23:59)")

    if len(data.matches) < 1 or len(data.matches) > 3:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Matches need to be 1 to 3")
    
    pool_set = {}
    team_set = {}
    for match in data.matches:
        if match.court_number < 1 or match.court_number > 10:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Court need to be 1 to 10")
        
        if match.team1_id == match.team2_id:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="One teams in the match")
        
        if match.court_number in pool_set:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Two matchs in the same pool")
        pool_set.add(match.court_number)

        if match.team2_id in team_set or match.team1_id in team_set:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="A team is playing twice in the same time")
        team_set.add(match.team1_id)
        team_set.add(match.team2_id)

    event = Event(
                event_date = data.event_date,
                event_time = data.event_time
            )

    db.add(event)

    for match in data.matches:
        team1 = db.query(Team).get(match.team1_id)
        team2 = db.query(Team).get(match.team2_id)

        if team1 is None or team2 is None:
            db.rollback()
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="One team not found")

        db.add(Match(
                court_number=match.court_number,
                team1=team1,
                team2=team2,
                event=event
            )
        )

    db.commit()
    db.refresh(event)
    return EventResponse.model_validate(event)



@router.put("/{event_id}", response_model=EventResponse)
def update_event(event_id: int, data: EventRequest, db: Session = Depends(get_db), _: str = Depends(get_current_admin)):
    """
    This function updates a event.
    
    param : event_id - The event's id.
    param : data - The event's informations.
    param : db - The session of database.
    param : _ - The client.
    return : Return the event updated.
    """
    if data.event_date < date.today():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Back to the future")
    
    if not re.match(r"^(?:[01]\d|2[0-3]):[0-5]\d$", data.event_time):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="event_time must be in HH:MM format (00:00–23:59)")

    if len(data.matches) < 1 or len(data.matches) > 3:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Matches need to be 1 to 3")
    
    pool_set = {}
    team_set = {}
    for match in data.matches:
        if match.court_number < 1 or match.court_number > 10:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Court need to be 1 to 10")
        
        if match.team1_id == match.team2_id:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="One teams in the match")
        
        if match.court_number in pool_set:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Two matchs in the same pool")
        pool_set.add(match.court_number)

        if match.team2_id in team_set or match.team1_id in team_set:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="A team is playing twice in the same time")
        team_set.add(match.team1_id)
        team_set.add(match.team2_id)
        
    event = db.query(Event).get(event_id)
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
    
    event.event_date = data.event_date
    event.event_time = data.event_time

    db.delete(event.matches)

    for match in data.matches:
        team1 = db.query(Team).get(match.team1_id)
        team2 = db.query(Team).get(match.team2_id)

        if team1 is None or team2 is None:
            db.rollback()
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="One team not found")

        db.add(Match(
                court_number=match.court_number,
                team1=team1,
                team2=team2,
                event=event
            )
        )

    db.commit()
    return EventResponse.model_validate(event)



@router.delete("/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_event(event_id: int, db: Session = Depends(get_db), _: str = Depends(get_current_admin)):
    """
    This function remove a event.

    param : event_id - The event's id.
    param : db - The session of database.
    param : _ - The client.
    return : Return no content
    """
    event = db.query(Event).get(event_id)
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
    
    for match in event.matches:
        if match.status != "A_VENIR":
            raise HTTPException(status_code=400, detail="Match over or cancel")

    db.delete(event)
    db.commit()

File: frontend-svelte/src/lib/services/event.ts
import api from './api';
import type { MatchInput } from './match';

/**
 * This interface represents a match inside an event.
 */
export interface EventMatchOutput {
  /**
   * Match identifier.
   */
  id: number;

  /**
   * Court number.
   */
  court_number: number;

  /**
   * Match status.
   */
  status: string;

  /**
   * This is the first team in the match.
   */ 
  team1_id: number

  /**
   * This is the second team in the match.
   */
  team2_id: number
}



/**
 * This interface represents an event.
 */
export interface EventOutput {
  /**
   * Event identifier.
   */
  id: number;

  /**
   * Event date.
   */
  event_date: string;

  /**
   * Event time.
   */
  event_time: string;

  /**
   * Event matches.
   */
  matches: EventMatchOutput[];
}



/**
 * This interface represents an event request.
 */
export interface EventInput {
  /**
   * Event date.
   */
  event_date: string;

  /**
   * Event time.
   */
  event_time: string;

  /**
   * Matches to create.
   */
  matches: MatchInput[];
}



export const eventsService = {

  /**
   * This function gets all events.
   * 
   * @param start_date The start date when to get the event (can be null).
   * @param end_date The end date when to get the event (can be null).
   * @param mine Filter to show only the event of the user.
   * @return Return all events.
   */
  getAllEvents(start_date?: Date, end_date?: Date, mine?: boolean) {
    return api.get<{ events: EventOutput[]; total: number }>('/events', {
      params: {
        start_date: start_date,
        end_date: end_date,
        mine: mine
      }
    });
  },



  /**
   * This function creates an event.
   * 
   * @param input Event informations.
   * @return Return the created event.
   */
  createEvent(input: EventInput) {
    return api.post<EventOutput>('/events', input);
  },



  /**
   * This function updates an event.
   * 
   * @param eventId Event identifier.
   * @param input Event informations.
   * @return Return the updated event.
   */
  updateEvent(eventId: number, input: EventInput) {
    return api.put<EventOutput>(`/events/${eventId}`, input);
  },



  /**
   * This function deletes an event.
   * 
   * @param eventId Event identifier.
   * @return Return no content.
   */
  deleteEvent(eventId: number) {
    return api.delete(`/events/${eventId}`);
  }
};
